File: BFS_DFS/network.cpp
#include <string>
#include <vector>
#include <iostream>
#include <cstring>
using namespace std;

bool check[200];

void dfs(int cur, int cnt, int n, vector<vector<int>> computers) {
	check[cur] = true;
	for (int next = 0; next < n; next++) {
		if (!check[next] && computers[cur][next] == 1) {
			dfs(next, cnt, n, computers);
		}
	}
}

int solution(int n, vector<vector<int>> computers) {
	int answer = 0;

	for (int i = 0; i < n; i++) {
		if (check[i]) continue;
		dfs(i, ++answer, n, computers);
	}

	return answer;
}

int main() {
	vector<vector<int>> graph = { {1, 1, 0}, {1, 1, 0}, {0, 0, 1} };
	vector < vector<int>> graph1 = { {1, 1, 0}, {1, 1, 1}, {0, 1, 1} };

	cout << solution(3, graph) << '\n';
	memset(check, false, sizeof(check));
	cout << solution(3, graph1) << '\n';

	return 0;
}

File: BFS_DFS/Python/network.py
def dfs(computers, v, visited, answer):
    # 현재 노드를 방문 처리
    visited[v] = True

    # 현재 노드와 연결된 다른 노드를 재귀적으로 방문
    for index, value in enumerate(computers[v]):
        if (not visited[index]) and value == 1:
            dfs(computers, index, visited, answer)


def solution(n, computers):
    answer = 0
    visited = [False] * n
    for v in range(n) :
        for index, value in enumerate(computers[v]):
            if (not visited[index]) and value == 1:
                dfs(computers, index, visited, answer)
                answer += 1
    return answer


if __name__ == "__main__" :
    n = int(input())
    computers = []
    for i in range(n) :
        computers.append(list(map(int, input().split())))
    print(solution(n, computers))
